### PR TITLE
gh-117578: Introduce _PyType_GetModuleByDef2 private function

### PR DESCRIPTION
**Pull request by neonene**
`_PyType_GetModuleByDefInPairs`(provisional name: **outdated**) is a private function that takes two types, runs `PyType_GetModuleByDef()` twice, and finally raises an error if any.

* Issue: gh-117578

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -4848,14 +4862,37 @@ PyType_GetModuleByDef(PyTypeObject *type, PyModuleDef *def)
...
+_PyType_GetModuleByDefInPairs(PyTypeObject *left, PyTypeObject *right,
+                              PyModuleDef *def)
+{
+    PyObject *module = type_get_module_by_def(left, def);
+    if (module == NULL) {
+        module = type_get_module_by_def(right, def);
+        if (module == NULL) {
+            PyErr_Format(
+                PyExc_TypeError,
+                "PyType_GetModuleByDef: No superclass of '%s' and '%s' has"
```
Comment (erlend-aasland) [inline Objects/typeobject.c]:
```suggestion
                "PyType_GetModuleByDef: No superclass of '%s' or '%s' has"
```

File: Objects/typeobject.c [hunk 2]
```diff
@@ -4848,14 +4862,37 @@ PyType_GetModuleByDef(PyTypeObject *type, PyModuleDef *def)
...
             PyExc_TypeError,
             "PyType_GetModuleByDef: No superclass of '%s' has the given module",
             type->tp_name);
     }
-    return res;
+    return module;
+}
+
+PyObject *
+_PyType_GetModuleByDefInPairs(PyTypeObject *left, PyTypeObject *right,
```
Comment (erlend-aasland) [inline Objects/typeobject.c]: I'm not sure about the naming. How about `_PyType_GetModuleByDefPair`, or perhaps even `_PyType_GetModuleByDef2`?

File: Objects/typeobject.c [hunk 3]
```diff
@@ -4888,7 +4888,7 @@ _PyType_GetModuleByDef2(PyTypeObject *left, PyTypeObject *right,
         if (module == NULL) {
             PyErr_Format(
                 PyExc_TypeError,
-                "PyType_GetModuleByDef: No superclass of '%s' and '%s' has"
+                "PyType_GetModuleByDef: No superclass of '%s' or '%s' has"
```
Comment (erlend-aasland) [inline Objects/typeobject.c]: Oh, I remember now the conjunction _nor_ should be used for negatives![^1] 

```suggestion
                "PyType_GetModuleByDef: No superclass of '%s' nor '%s' has"
```

[^1]: quoting my dictionary: nor -- used before the second or further of two or more alternatives (the first being introduced by a negative such as ‘neither’ or ‘not’) to indicate that they are each untrue or each do not happen

**Comment by encukou** [comment]
Please don't change the documented behaviour of `PyType_GetModuleByDef`. It should return the first match.
That said, if you have several different modules with the same def in a MRO, things are probably not going the way you wanted. The distinction between the first/last matching entry would make more sense if you want the base type itself, not its `ht_module`.

File: Objects/typeobject.c [hunk 4]
```diff
@@ -4840,9 +4853,10 @@ PyType_GetModuleByDef(PyTypeObject *type, PyModuleDef *def)
     // mro_invoke() ensures that the type MRO cannot be empty, so we don't have
     // to check i < PyTuple_GET_SIZE(mro) at the first loop iteration.
     assert(PyTuple_GET_SIZE(mro) >= 1);
+    assert(PyTuple_GET_ITEM(mro, 0) == (PyObject *)type);
```
Comment (encukou) [inline Objects/typeobject.c]: Let's add a comment about why the loop starts at one.

The comment about "first loop iteration" should have been removed when Victor accidentally changed the MRO walk from a do-while loop to a for; let's remove it.
```suggestion
    // mro_invoke() ensures that the type MRO cannot be empty.
    assert(PyTuple_GET_SIZE(mro) >= 1);
    // Also, the first item in the MRO is the type itself, which
    // we already checked above. We skip it in the loop.
    assert(PyTuple_GET_ITEM(mro, 0) == (PyObject *)type);
```

File: Objects/typeobject.c [hunk 5]
```diff
@@ -4825,11 +4825,24 @@ PyType_GetModuleState(PyTypeObject *type)
...
+        // type_ready_mro() ensures that no heap type is
+        // contained in a static type MRO.
+        return NULL;
+    }
+    else {
+        PyHeapTypeObject *ht = (PyHeapTypeObject*)type;
+        PyObject *module = ht->ht_module;
+        if (module && _PyModule_GetDef(module) == def) {
+            return module;
+        }
```
Comment (encukou) [inline Objects/typeobject.c:4843]: (Does this special case actually speed things up?
`lookup_tp_mro`, `PyTuple_GET_SIZE` and `PyTuple_GET_ITEM` are rather fast, so I wouldn't be surprised if starting the loop at 0 and doing the check there was wast enough.)

Comment (neonene) [reply Objects/typeobject.c:4843]: I feel that a static type should be rejected here at least for the pair-args function.  Also, I've found a similar attempt that says 5% difference on `lru_cache()`: #25393.

Maybe the initial check makes more sense with GIL disabled?

Comment (encukou) [reply Objects/typeobject.c:4843]: Makes sense.

**Review by encukou — APPROVED** [review]
LGTM, thank you!
As for tests, this should be covered by the `decimal` test suite, but if you want to write a `GetModuleByDef` test please ping me for a review.

**Comment by mdboom** [comment]
@neonene: We should have measured the performance impact of this at the time, but it came up recently in conversation with @markshannon, so I decided to measure it.

While overall on the benchmarking suite it's kind of inconclusive (1% faster, which is below the usual noise threshold), for the one benchmark that uses `_PyType_GetModuleByDef` heavily (via that `fraction` extension), it saw a 5% improvement, as well as a 3% decrease in memory usage.  [Results here](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20240425-3.13.0a6+-2c45148).

**Pull request merged**